src/validation.cpp

```diff
@@ -1060,28 +1176,93 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+    return MempoolAcceptResult(*ws.m_ptx, std::move(ws.m_replaced_transactions), ws.m_base_fees);
+}
+
+std::vector<MempoolAcceptResult> MemPoolAccept::AcceptMultipleTransactions(std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+    std::vector<Workspace> workspaces{};
+    const int package_size = txns.size();
+    workspaces.reserve(package_size);
+    std::transform(txns.begin(), txns.end(), std::back_inserter(workspaces), [](CTransactionRef& tx) {
+        return Workspace(tx);
+    });
+    std::vector<MempoolAcceptResult> results;
+    results.reserve(package_size);
+
+    LOCK(m_pool.cs);
+    // Do all PreChecks first and fail fast to avoid running expensive script checks when unnecessary.
+    for (Workspace& ws : workspaces) {
+        if (!PreChecks(args, ws)) {
+            // Exit early to avoid doing pointless work. Return results in the same order as input txns.
+            const auto failed_or_unfinished = [&, failed_ptx = ws.m_ptx](Workspace& ws) {
+                                              return ws.m_ptx == failed_ptx
+                                              ? MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ true)
+                                              : MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ false);
+            };
+            std::transform(workspaces.begin(), workspaces.end(), std::back_inserter(results), failed_or_unfinished);
+            return results;
+        }
+        m_view.PackageAddTransaction(ws.m_ptx);
+    }
+
+    // Now that we have verified all inputs are available and there are no conflicts in the package,
+    // clear the temporary coins (m_temp_added and m_temp_spent), otherwise script checks will error
+    // on coins that are spent within the package.
+    m_view.ClearTemporaryCoins();
+
+    for (Workspace& ws : workspaces) {
+        PrecomputedTransactionData txdata;
+        if (!PolicyScriptChecks(args, ws, txdata)) {
+            // Exit early to avoid doing pointless work. Return results in the same order as input txns.
+            CTransactionRef failed_ptx = ws.m_ptx;
+            const auto failed_or_unfinished = [&failed_ptx](Workspace& ws) {
+                                              return ws.m_ptx == failed_ptx
+                                              ? MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ true)
+                                              : MempoolAcceptResult(*ws.m_ptx, ws.m_state, /* finished */ false);
+            };
+            auto it_curr = std::find_if(workspaces.begin(), workspaces.end(),
+                                        [& failed_ptx](Workspace& ws) { return ws.m_ptx == failed_ptx; });
+            // When test_accept=true, transactions that pass PolicyScriptChecks are valid because there are
+            // no further mempool checks (passing PolicyScriptChecks implies passing ConsensusScriptChecks).
```

Comment:
> I think you're forgetting the mempool size check in Finalize

We have a feerate check against the mempool minimum fee in `CheckFeeRate()` (in `PreChecks()`). The trim to size is done after submitting to mempool just in case it wasn't accurate (and I suspect the mempool lock was perhaps not held the whole time in the past, so it was possible for the mempool to grow in the time we were doing script checks?). On master, a test_accept for one transaction stops short of calling `Finalize()` and doesn't do this check either. 

I prefer using `CheckFeeRate()` for a test_accept; there's no need to submit to mempool, update everything, and then remove it again and re-update everything, just to check size limits. Since we would release the lock at the end of the testmempoolaccept call, it's possible for the mempool to grow before the user submits it anyway, so it's not like we're guaranteed an accurate result by doing this.

Reply:
> Beyond the fact that policy flags are a superset of the consensus ones, you could also recall that's running `ConsensusScriptChecks` isn't worthy here as we don't try to cache script results.

Yep, I agree! See also [this discussion](https://github.com/bitcoin/bitcoin/pull/20833#discussion_r552684985).

Reply:
I've noted in the release notes that the true mempool minimum fee may not be met in the test_accept

Perhaps in a separate PR (which would be relevant even for non-package-accept) we could figure out how to get a more accurate mempool minimum fee without submitting to it :)